File: globalLMI.m
function [ PgMax ] = globalLMI( physics, vref )
%UNTITLED Summary of this function goes here
%   Detailed explanation goes here

    m = physics.m;
    g = physics.g;
    kv = physics.kv;
    k = physics.k;
    FS = physics.muS*m*g;
    FC = physics.muC*m*g;

    % Define the decision variables
    Pg = sdpvar(2);
    tau1 = sdpvar(1);
    tau2 = sdpvar(1);
    tau3 = sdpvar(1);
    tau4 = sdpvar(1);
    tau5 = sdpvar(1);
    
    eta = sdpvar(1);
    
    % Creation of Theta
    A = [-kv/m -k/m; 1 0];
    B = [-1/m; 0];
    Fnl = @(thetaDot) friction(thetaDot, physics) - kv*thetaDot;
    D = [A B -Fnl(vref)*B];
    F = eye(2, 4);
    
    Theta = D'*Pg*F + F'*Pg*D;
    
    % Creation of Pi matrices
    pi1 = [1 0 0 0];
    pi3 = [0 0 1 0];
    pi4 = [0 0 0 1];
    
    Pi0 = (pi4'*pi4) - F'*Pg*F;
    Pi1 = pi3'*pi3 - FS^2*(pi4'*pi4);
    Pi2 = -pi3'*pi3 + FC^2*(pi4'*pi4);
    Pi3 = -(pi1 + vref*pi4)'*pi3;
    Pi3 = Pi3' + Pi3;
    Pi4 = (pi + vref*pi4)'*(pi + vref*pi4);
    % Creation of ThetaBar
    
    tau0List = linspace(0, -2*max(real(eig(A))), 30);
    
    etaMax = 0;
    PgMax = nan;
    for tau0 = tau0List

        ThetaBar = Theta - tau0*Pi0 - tau1*Pi1 - tau2*Pi2 - tau3*Pi3;
        ThetaBar2 = Theta - tau0*Pi0 - tau5*Pi1 - tau4*Pi4;
        
        % Soving the optimization problem    
        Constraints = [(Pg >= eta*eye(2)):'Positivity',...
            (tau1 >= 0):'S-variable F_{max}',...
            (tau2 >= 0):'S-variable F_{min}'...
            (tau3 >= 0):'S-variable sign'...
            (tau5 >= 0):'S-variable F_{max}',...
            (ThetaBar <= -1e-5):'Negativity',...
            (ThetaBar2 <= -1e-5):'Negativity'];

        warning('off', 'YALMIP:strict');
        option = sdpsettings('verbose', 0, 'solver', 'mosek', 'cachesolvers', 1);
        diagnostics = optimize(Constraints, -eta, option);

        if diagnostics.problem == 0
            if value(eta) >= etaMax
                etaMax = value(eta);
                PgMax = value(Pg);
            end
        end
    end

end

File: global_LMI.py
from friction import friction 

import numpy as np
import cvxpy as cp
import math

def global_LMI(SS, model, vref, tol_eps=1e-6):
    FS = model.muS * model.m * model.g
    FC = model.muC * model.m * model.g

    Pg = cp.Variable((2, 2), symmetric=True)
    tau0 = None                 
    tau1 = cp.Variable(nonneg=True)
    tau2 = cp.Variable(nonneg=True)
    tau3 = cp.Variable(nonneg=True)
    tau4 = cp.Variable()        
    tau5 = cp.Variable(nonneg=True)
    eta = cp.Variable()

    A = SS.A
    B = SS.B


    Fnl = friction(vref, model) - model.kv * vref
    D = np.hstack([A, B, -Fnl * B])
    F = np.eye(2, 4)
    #F = np.hstack([np.eye(2), np.zeros((2,2))])  


    Theta = D.T @ Pg @ F + F.T @ Pg @ D
   

    pi1 = np.array([[1, 0, 0, 0]])
    pi3 = np.array([[0, 0, 1, 0]])
    pi4 = np.array([[0, 0, 0, 1]])

    Pi0 = pi4.T @ pi4 - F.T @ Pg @ F
    Pi1 = pi3.T @ pi3 - (FS**2) * (pi4.T @ pi4)
    Pi2 = (FC**2) * (pi4.T @ pi4) - pi3.T @ pi3
    tmp = (pi1 + vref * pi4).T @ pi3      
    Pi3 = -(tmp + tmp.T)
    Pi4 = (pi1 + vref * pi4).T @ (pi1 + vref * pi4)
    #Pi4 = (math.pi + vref * pi4).T @ (math.pi + vref * pi4)

    eigA = np.linalg.eigvals(A)
    tau0_max = -2 * np.max(np.real(eigA))
    tau0_list = np.linspace(0, tau0_max, 30)

    best_Pg = None
    best_eta = -1e12

    eps = max(tol_eps, 1e-10)


    for tau0 in tau0_list:
        ThetaBar = Theta - tau0 * Pi0 - tau1 * Pi1 - tau2 * Pi2 - tau3 * Pi3
        ThetaBar2 = Theta - tau0 * Pi0 - tau5 * Pi1 - tau4 * Pi4

        n = Pg.shape[0]
        constraints = [
            Pg - eta * np.eye(2) >> 0,
            tau1 >= 0,
            tau2 >= 0,
            tau3 >= 0,
            # tau4 unconstrained (do NOT add >=0)
            tau5 >= 0,
            ThetaBar << -eps * np.eye(ThetaBar.shape[0]),
            ThetaBar2 << -eps * np.eye(ThetaBar2.shape[0]),
            Pg >> 1e-8 * np.eye(2)   # ensure strictly positive definite
        ]

        prob = cp.Problem(cp.Maximize(eta), constraints)
        prob.solve(solver=cp.SCS, verbose=True)

        if prob.status in ["optimal", "optimal_inaccurate"]:
            val = eta.value if eta.value is not None else -1e12
            if val > best_eta:
                best_eta = float(val)
                best_Pg = Pg.value.copy()

    return best_Pg